README.md
Akami [![Build Status](http://travis-ci.org/savonrb/akami.png)](http://travis-ci.org/savonrb/akami)
=====

Building Web Service Security. This fork is to allow Nonce to be signed and Created to be digested.


Installation
------------

Akami is available through [Rubygems](http://rubygems.org/gems/akami) and can be installed via:

```
$ gem install akami
```


Getting started
---------------

``` ruby
wsse = Akami.wsse
```

Set the credentials for `wsse:UsernameToken` basic auth:

``` ruby
wsse.credentials "username", "password"
```

Set the credentials for `wsse:UsernameToken` digest auth:

``` ruby
wsse.credentials "username", "password", :digest
```

Enable `wsu:Timestamp` headers. `wsu:Created` is automatically set to `Time.now`
and `wsu:Expires` is set to `Time.now + 60`:

``` ruby
wsse.timestamp = true
```

Manually specify the values for `wsu:Created` and `wsu:Expires`:

``` ruby
wsse.created_at = Time.now
wsse.expires_at = Time.now + 60
```

Akami is based on an autovivificating Hash. So if you need to add custom tags, you can add them.

``` ruby
wsse["wsse:Security"]["wsse:UsernameToken"] = { "Organization" => "ACME" }
```

When generating the XML for the request, this Hash will be merged with another Hash containing
all the default tags and values.  
This way you might digg into some code, but then you can even overwrite the default values.

``` ruby
wsse.to_xml
```


CHANGELOG.md
## 1.2.1
* Allow nonce to be public key signed.
* Allow to digest created_at field.

## master

* Fix: [#2](https://github.com/savonrb/akami/pull/2) Fixes related to WS-Security,
  UserToken authentication.

## 1.2.0 (2012-06-28)

* Fix: Lowered the version of Nokogiri required to use Akami.

## 1.1.0 (2012-06-06)

* Feature: [#3](https://github.com/savonrb/akami/pull/3) - WSSE signing.

## 1.0.0 (2011-07-03)

* Initial version extracted from the [Savon](http://rubygems.org/gems/savon) library.


spec/spec_helper.rb
require "bundler"
Bundler.require :default, :development
def resource_path(path)
  "#{File.dirname(__FILE__)}/resources/#{path}"
end

def resource_read(path)
  File.read("#{resource_path(path)}")
end

spec/akami/wsse_spec.rb
require "spec_helper"
require "base64"

describe Akami do
  let(:wsse) { Akami.wsse }

  it "contains the namespace for WS Security Secext" do
    Akami::WSSE::WSE_NAMESPACE.should ==
      "http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd"
  end

  it "contains the namespace for WS Security Utility" do
    Akami::WSSE::WSU_NAMESPACE.should ==
      "http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd"
  end

  it "contains the namespace for the PasswordText type" do
    Akami::WSSE::PASSWORD_TEXT_URI.should ==
      "http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-username-token-profile-1.0#PasswordText"
  end

  it "contains the namespace for the PasswordDigest type" do
    Akami::WSSE::PASSWORD_DIGEST_URI.should ==
      "http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-username-token-profile-1.0#PasswordDigest"
  end

  describe "#credentials" do
    it "sets the username" do
      wsse.credentials "username", "password"
      wsse.username.should == "username"
    end

    it "sets the password" do
      wsse.credentials "username", "password"
      wsse.password.should == "password"
    end

    it "defaults to set digest to false" do
      wsse.credentials "username", "password"
      wsse.should_not be_digest
    end

    it "sets digest to true if specified" do
      wsse.credentials "username", "password", :digest
      wsse.should be_digest
    end
  end

  describe "#username" do
    it "sets the username" do
      wsse.username = "username"
      wsse.username.should == "username"
    end
  end

  describe "#password" do
    it "sets the password" do
      wsse.password = "password"
      wsse.password.should == "password"
    end
  end

  describe "#digest" do
    it "defaults to false" do
      wsse.should_not be_digest
    end

    it "specifies whether to use digest auth" do
      wsse.digest = true
      wsse.should be_digest
    end
  end

  describe "#to_xml" do
    context "with no credentials" do
      it "returns an empty String" do
        wsse.to_xml.should == ""
      end
    end

    context "with only a username" do
      before { wsse.username = "username" }

      it "returns an empty String" do
        wsse.to_xml.should == ""
      end
    end

    context "with only a password" do
      before { wsse.password = "password" }

      it "returns an empty String" do
        wsse.to_xml.should == ""
      end
    end

    context "with credentials" do
      before { wsse.credentials "username", "password" }

      it "contains a wsse:Security tag" do
        namespace = "http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd"
        wsse.to_xml.should include("<wsse:Security xmlns:wsse=\"#{namespace}\">")
      end

      it "contains a wsu:Id attribute" do
        wsse.to_xml.should include('<wsse:UsernameToken wsu:Id="UsernameToken-1"')
      end

      it "increments the wsu:Id attribute count" do
        wsse.to_xml.should include('<wsse:UsernameToken wsu:Id="UsernameToken-1"')
        wsse.to_xml.should include('<wsse:UsernameToken wsu:Id="UsernameToken-2"')
      end

      it "contains the WSE and WSU namespaces" do
        wsse.to_xml.should include(Akami::WSSE::WSE_NAMESPACE, Akami::WSSE::WSU_NAMESPACE)
      end

      it "contains the username and password" do
        wsse.to_xml.should include("username", "password")
      end

      it "does not contain a wsse:Nonce tag" do
        wsse.to_xml.should_not match(/<wsse:Nonce>.*<\/wsse:Nonce>/)
      end

      it "does not contain a wsu:Created tag" do
        wsse.to_xml.should_not match(/<wsu:Created>.*<\/wsu:Created>/)
      end

      it "contains the PasswordText type attribute" do
        wsse.to_xml.should include(Akami::WSSE::PASSWORD_TEXT_URI)
      end
    end

    context "with credentials and digest auth" do
      before { wsse.credentials "username", "password", :digest }

      it "contains the WSE and WSU namespaces" do
        wsse.to_xml.should include(Akami::WSSE::WSE_NAMESPACE, Akami::WSSE::WSU_NAMESPACE)
      end

      it "contains the username" do
        wsse.to_xml.should include("username")
      end

      it "does not contain the (original) password" do
        wsse.to_xml.should_not include("password")
      end

      it "contains a wsse:Nonce tag" do
        wsse.to_xml.should match(/<wsse:Nonce>[^<]+<\/wsse:Nonce>/)
      end

      it "contains a wsu:Created tag" do
        created_at = Time.now
        Timecop.freeze created_at do
          wsse.to_xml.should include("<wsu:Created>#{created_at.utc.xmlschema}</wsu:Created>")
        end
      end

      it "contains the PasswordDigest type attribute" do
        wsse.to_xml.should include(Akami::WSSE::PASSWORD_DIGEST_URI)
      end

      it "should reset the nonce every time" do
        created_at = Time.now
        Timecop.freeze created_at do
          nonce_regexp = /<wsse:Nonce>([^<]+)<\/wsse:Nonce>/
          nonce_first = Base64.decode64(nonce_regexp.match(wsse.to_xml)[1])
          nonce_second = Base64.decode64(nonce_regexp.match(wsse.to_xml)[1])
          nonce_first.should_not == nonce_second
        end
      end

      it "has contains a properly hashed password" do
        xml_header = Nokogiri::XML(wsse.to_xml)
        xml_header.remove_namespaces!
        nonce = Base64.decode64(xml_header.xpath('//Nonce').first.content)
        created_at = xml_header.xpath('//Created').first.content
        password_hash = Base64.decode64(xml_header.xpath('//Password').first.content)
        password_hash.should == Digest::SHA1.digest((nonce + created_at + "password"))
      end
    end

    context "with #timestamp set to true" do
      before { wsse.timestamp = true }

      it "contains a wsse:Timestamp node" do
        wsse.to_xml.should include('<wsu:Timestamp wsu:Id="Timestamp-1" ' +
          'xmlns:wsu="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd">')
      end

      it "contains a wsu:Created node defaulting to Time.now" do
        created_at = Time.now
        Timecop.freeze created_at do
          wsse.to_xml.should include("<wsu:Created>#{created_at.utc.xmlschema}</wsu:Created>")
        end
      end

      it "contains a wsu:Expires node defaulting to Time.now + 60 seconds" do
        created_at = Time.now
        Timecop.freeze created_at do
          wsse.to_xml.should include("<wsu:Expires>#{(created_at + 60).utc.xmlschema}</wsu:Expires>")
        end
      end
    end

    context "with #created_at" do
      before { wsse.created_at = Time.now + 86400 }

      it "contains a wsu:Created node with the given time" do
        wsse.to_xml.should include("<wsu:Created>#{wsse.created_at.utc.xmlschema}</wsu:Created>")
      end

      it "contains a wsu:Expires node set to #created_at + 60 seconds" do
        wsse.to_xml.should include("<wsu:Expires>#{(wsse.created_at + 60).utc.xmlschema}</wsu:Expires>")
      end
    end

    context "with #expires_at" do
      before { wsse.expires_at = Time.now + 86400 }

      it "contains a wsu:Created node defaulting to Time.now" do
        created_at = Time.now
        Timecop.freeze created_at do
          wsse.to_xml.should include("<wsu:Created>#{created_at.utc.xmlschema}</wsu:Created>")
        end
      end

      it "contains a wsu:Expires node set to the given time" do
        wsse.to_xml.should include("<wsu:Expires>#{wsse.expires_at.utc.xmlschema}</wsu:Expires>")
      end
    end

    context "whith credentials and timestamp" do
      before do
        wsse.credentials "username", "password"
        wsse.timestamp = true
      end

      it "contains a wsu:Created node" do
        wsse.to_xml.should include("<wsu:Created>")
      end

      it "contains a wsu:Expires node" do
        wsse.to_xml.should include("<wsu:Expires>")
      end

      it "contains the username and password" do
        wsse.to_xml.should include("username", "password")
      end
    end

    context "with credentials and digest auth and ciphered elements " do
      before {
        wsse.credentials "username", "password", :ciphered_digest
        wsse.digest_fields= [:created_at]
        certificate= OpenSSL::X509::Certificate.new resource_read('akami.crt')
        wsse.cipher_nonce OpenSSL::PKey::RSA.new(certificate.public_key)
      }

      it "contains the WSE and WSU namespaces" do
        wsse.to_xml.should include(Akami::WSSE::WSE_NAMESPACE, Akami::WSSE::WSU_NAMESPACE)
      end

      it "contains the username" do
        wsse.to_xml.should include("username")
      end

      it "does not contain the (original) password" do
        wsse.to_xml.should_not include("password")
      end

      it "contains a wsse:Nonce tag" do
        wsse.to_xml.should match(/<wsse:Nonce>[^<]+<\/wsse:Nonce>/)
      end

      it "does not contain the plain created_at value" do
        created_at = Time.now
        Timecop.freeze created_at do
          wsse.to_xml.should_not include("<wsu:Created>#{created_at.utc.xmlschema}</wsu:Created>")
        end
      end

      it "contains a properly hashed wsu:Created tag" do
        created_at = Time.now
        Timecop.freeze created_at do
          xml_header = Nokogiri::XML(wsse.to_xml)
          xml_header.remove_namespaces!
          created_at_hash = Base64.decode64(xml_header.xpath('//Created').first.content)
          created_at_hash.should == Digest::SHA1.digest(created_at.utc.xmlschema)
        end
      end

      it "contains the PasswordDigest type attribute" do
        wsse.to_xml.should include(Akami::WSSE::PASSWORD_DIGEST_URI)
      end

      it "should reset the nonce every time" do
        created_at = Time.now
        Timecop.freeze created_at do
          nonce_regexp = /<wsse:Nonce>([^<]+)<\/wsse:Nonce>/
          nonce_first = Base64.decode64(nonce_regexp.match(wsse.to_xml)[1])
          nonce_second = Base64.decode64(nonce_regexp.match(wsse.to_xml)[1])
          nonce_first.should_not == nonce_second
        end
      end

      it "has contains a properly hashed password" do
        created_at = Time.now
        Timecop.freeze created_at do
          xml_header = Nokogiri::XML(wsse.to_xml)
          xml_header.remove_namespaces!
          nonce= Base64.decode64(xml_header.xpath('//Nonce').first.content)
          pki= OpenSSL::PKey::RSA.new(resource_read('akami.key'), 'test')
          nonce= pki.private_decrypt(nonce)
          password_hash = Base64.decode64(xml_header.xpath('//Password').first.content)
          password_hash.should == Digest::SHA1.digest((nonce + created_at.utc.xmlschema + "password"))
        end
      end
    end
  end

end


lib/akami/wsse.rb
require "base64"
require "digest/sha1"
require "akami/core_ext/hash"
require "akami/xpath_helper"
require "akami/c14n_helper"
require "time"
require "gyoku"

require "akami/wsse/verify_signature"
require "akami/wsse/signature"

module Akami

  # = Akami::WSSE
  #
  # Building Web Service Security.
  class WSSE

    # Namespace for WS Security Secext.
    WSE_NAMESPACE = "http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd"

    # Namespace for WS Security Utility.
    WSU_NAMESPACE = "http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd"

    # PasswordText URI.
    PASSWORD_TEXT_URI = "http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-username-token-profile-1.0#PasswordText"

    # PasswordDigest URI.
    PASSWORD_DIGEST_URI = "http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-username-token-profile-1.0#PasswordDigest"

    # Returns a value from the WSSE Hash.
    def [](key)
      hash[key]
    end

    # Sets a value on the WSSE Hash.
    def []=(key, value)
      hash[key] = value
    end

    # Sets authentication credentials for a wsse:UsernameToken header.
    # Also accepts whether to use WSSE digest authentication.
    def credentials(username, password, digest = false)
      self.username = username
      self.password = password
      self.digest = digest
    end

    attr_accessor :username, :password, :created_at, :expires_at, :signature, :verify_response, :digest_fields

    # indicate to cipher the nonce
    def cipher_nonce pki
      @pki= pki
    end

    def sign_with=(klass)
      @signature = klass
    end

    def signature?
      !!@signature
    end

    # Returns whether to use WSSE digest. Defaults to +false+.
    def digest?
      !!@digest
    end

    attr_writer :digest

    # Returns whether to generate a wsse:UsernameToken header.
    def username_token?
      username && password
    end

    # Returns whether to generate a wsu:Timestamp header.
    def timestamp?
      created_at || expires_at || @wsu_timestamp
    end

    # Sets whether to generate a wsu:Timestamp header.
    def timestamp=(timestamp)
      @wsu_timestamp = timestamp
    end

    # Hook for Soap::XML that allows us to add attributes to the env:Body tag
    def body_attributes
      if signature?
        signature.body_attributes
      else
        {}
      end
    end

    # Returns the XML for a WSSE header.
    def to_xml
      if signature? and signature.have_document?
        Gyoku.xml wsse_signature.merge!(hash)
      elsif username_token? && timestamp?
        Gyoku.xml wsse_username_token.merge!(wsu_timestamp) {
          |key, v1, v2| v1.merge!(v2) {
            |key, v1, v2| v1.merge!(v2)
          }
        }
      elsif username_token?
        Gyoku.xml wsse_username_token.merge!(hash)
      elsif timestamp?
        Gyoku.xml wsu_timestamp.merge!(hash)
      else
        ""
      end
    end

  private

    # Returns a Hash containing wsse:UsernameToken details.
    def wsse_username_token
      if digest?
#        nonce= self.nonce
        tstmp= timestamp
        if defined?(@digest_fields) and @digest_fields.include?(:created_at)
          tstmp= digest_created_at
        end
        _nonce= nonce
        if defined?(@pki) and !@pki.nil?
          _nonce= @pki.public_encrypt(_nonce)
        end
        token = security_hash :wsse, "UsernameToken",
          "wsse:Username" => username,
          "wsse:Nonce" => Base64.encode64(_nonce),
          "wsu:Created" => tstmp,
          "wsse:Password" => digest_password,
          :attributes! => { "wsse:Password" => { "Type" => PASSWORD_DIGEST_URI } }
        # clear the nonce after each use
        @nonce = nil
      else
        token = security_hash :wsse, "UsernameToken",
          "wsse:Username" => username,
          "wsse:Password" => password,
          :attributes! => { "wsse:Password" => { "Type" => PASSWORD_TEXT_URI } }
      end
      token
    end

    def wsse_signature
      signature_hash = signature.to_token

      # First key/value is tag/hash
      tag, hash = signature_hash.shift

      security_hash nil, tag, hash, signature_hash
    end

    # Returns a Hash containing wsu:Timestamp details.
    def wsu_timestamp
      security_hash :wsu, "Timestamp",
        "wsu:Created" => (created_at || Time.now).utc.xmlschema,
        "wsu:Expires" => (expires_at || (created_at || Time.now) + 60).utc.xmlschema
    end

    # Returns a Hash containing wsse/wsu Security details for a given
    # +namespace+, +tag+ and +hash+.
    def security_hash(namespace, tag, hash, extra_info = {})
      key = [namespace, tag].compact.join(":")

      sec_hash = {
        "wsse:Security" => {
          key => hash
        },
        :attributes! => { "wsse:Security" => { "xmlns:wsse" => WSE_NAMESPACE } }
      }

      unless extra_info.empty?
        sec_hash["wsse:Security"].merge!(extra_info)
      end

      if signature?
        sec_hash[:attributes!].merge!("soapenv:mustUnderstand" => "1")
      else
        sec_hash["wsse:Security"].merge!(:attributes! => { key => { "wsu:Id" => "#{tag}-#{count}", "xmlns:wsu" => WSU_NAMESPACE } })
      end

      sec_hash
    end

    # Returns the WSSE password, encrypted for digest authentication.
    def digest_password
      token = nonce + timestamp + password
      Base64.encode64(Digest::SHA1.digest(token)).chomp!
    end

    def digest_created_at
      Base64.encode64(Digest::SHA1.digest(timestamp)).chomp!
    end

    # Returns a WSSE nonce.
    def nonce
      @nonce ||= Digest::SHA1.hexdigest random_string + timestamp
    end

    # Returns a random String of 100 characters.
    def random_string
      (0...100).map { ("a".."z").to_a[rand(26)] }.join
    end

    # Returns a WSSE timestamp.
    def timestamp
      @timestamp ||= Time.now.utc.xmlschema
    end

    # Returns a new number with every call.
    def count
      @count ||= 0
      @count += 1
    end

    # Returns a memoized and autovivificating Hash.
    def hash
      @hash ||= Hash.new { |h, k| h[k] = Hash.new(&h.default_proc) }
    end

  end
end
